### ISSUE TITLE
Case bindings in CaseStatement and CaseExpression should allow empty tuple.

### ISSUE THREAD

**Issue by richardlford** [dafny-lang/dafny#32]
The CaseStatement and CaseExpression grammar currently has the following
(in Dafny.atg):

```
  "case"                      (. x = t; .)
  ( ...
  | "("
      CasePattern<out pat>        (. arguments.Add(pat); .)
      { "," CasePattern<out pat>  (. arguments.Add(pat); .)   }
    ")"
```

This does not allow to match an empty tuple. It should be:

```
  "case"                      (. x = t; .)
  ( ...
  | "("
      [ 
         CasePattern<out pat>        (. arguments.Add(pat); .)
         { "," CasePattern<out pat>  (. arguments.Add(pat); .)  }
      ]
    ")"
```

**Comment by jaylorch** [comment]
Could you give an example of how this would be used? I can't figure out what kind of Dafny expression can match on both an empty tuple and something else.

**Comment by wilcoxjay** [comment]
I don't think the claim is that this reduces expressiveness (because indeed, anything that could be matched against `()` might as well be matched against `_`), but rather that it is a "sharp corner" in the language design. Patterns support tuples, so they should support all tuples, even zero-length ones.

**Comment by davidcok** [comment]
Fixed by PR#681
